NFTL Disk Driver

LinderPi opened the pull request:
Zephyr has currently no flash translation layer implemented. This pull request adds the basic possibility to use Dhara as FTL in Zephyr. It is implemented as disk driver that can be placed between a filesystem and a flash driver. The FTL uses API extensions for bad block management, which are also faked on the flash simulator to facilitate tests.

The [Dhara repository](https://github.com/endresshauser-lp/dhara/tree/zephyr) shall be moved to the Zephyr organisation as soon as the module is ready for that (#99908, #100834).

de-nordic commented on drivers/flash/flash_simulator.c:
```diff
@@ -380,6 +381,35 @@ flash_sim_get_parameters(const struct device *dev)
+	case NAND_FLASH_MARK_BAD_BLOCK: {
+		/* Blocks cannot be marked bad on flash simulator */
+		break;
+	}
```
I think we should also return -ENOTSUP here, we should describe return values for the NAND_FLASH_MARK_BAD_BLOCK and code calling the op should properly address the ENTOSUP.

de-nordic replied:
I am also not sure whether Flash Simulator, in the current form, is able to properly pretend NAND devices.
I think we should alter it to be able to use it for testing FTL on the CI, which also means that we may have to add ability to actually support bad blocks, as well as ecc or user bits.
I am not sure whether this is job for this pr, nevertheless we will not be able to reliably test ftl without it.

de-nordic commented on include/zephyr/drivers/flash/nand_flash_api_ex.h:
```diff
@@ -0,0 +1,33 @@
+	/**
+	 * Marks a block as bad. As input it takes the block index (size_t *). There is no output.
+	 */
+	NAND_FLASH_MARK_BAD_BLOCK,
```
It is not vendor base, that is something we should standardize.
I think that we also should drop NAND_, as not only NAND may have such features.

de-nordic commented on modules/dhara/Kconfig:
```diff
@@ -0,0 +1,25 @@
+	default y
+	depends on DT_HAS_ZEPHYR_FTL_DHARA_ENABLED
+	depends on DISK_DRIVERS
+	select FLASH_EX_OP_ENABLED
```
if DISK_DRIVER_FLASH

de-nordic commented on drivers/flash/flash_simulator.c:
```diff
@@ -380,6 +381,35 @@ flash_sim_get_parameters(const struct device *dev)
+	switch (code) {
+	case NAND_FLASH_IS_BAD_BLOCK: {
+		/* Block is never bad on flash simulator */
+		*(int *)out = 0;
```
GOOD = 0, BAD = !GOOD, name values.

de-nordic commented on drivers/disk/ftl_dhara.c:
```diff
@@ -0,0 +1,424 @@
+
+	ret = flash_ex_op(ctx->info.dev, NAND_FLASH_IS_BAD_BLOCK, (uintptr_t)&b, &is_bad);
+	if (ret == -ENOTSUP) {
+		LOG_INF("Checking bad block is not supported");
```
LOG_DBG as this is only needed on testing and evaluation of user system setup. Otherwise you will have a lot of these.

de-nordic commented on drivers/disk/ftl_dhara.c:
```diff
@@ -0,0 +1,424 @@
+		if (ctx->page_buffer[i] != flash_params->erase_value) {
+			return 0;
+		}
+	}
```
memset

de-nordic commented on drivers/disk/ftl_dhara.c:
```diff
@@ -0,0 +1,424 @@
+			},                                                                         \
+		.partition_id = DT_FIXED_PARTITION_ID(PARTITION_PHANDLE(n)),                       \
+		.partition_offset = DT_REG_ADDR(PARTITION_PHANDLE(n)),                             \
+		.partition_size = DT_REG_SIZE(PARTITION_PHANDLE(n)),                               \
```
Use FIXED_PARTITION macros here, and instead of ID, use the macro, FIXED_PARTITION, that directly takes pointer to `struct flash_area` object instance serving the selected partition.

LinderPi commented on drivers/disk/ftl_dhara.c:
```diff
@@ -0,0 +1,424 @@
+		LOG_INF("Checking bad block is not supported");
+		return 0;
+	} else if (ret != 0) {
+		LOG_ERR("Checking bad block failed with error %d", ret);
```
I added the block address and a info log for marked bad blocks. In future, this disk driver shall provide an additional health monitoring API to allow applications checking the available good blocks.

tpambor commented:
> I've little confused. Should we need to add ctx->partition_offset back to every flash access functions, just like ctx->offset of flashdisk.c did?

Yes, you are right. I have fixed it.

tpambor commented:
- Updated to latest disk driver API including erase/trim support
- Added support for disk_access tests using native_sim
- Moved KConfig to disk drivers

de-nordic commented on drivers/disk/ftl_dhara.c:
```diff
@@ -0,0 +1,508 @@
+
+	for (uint32_t i = 0; i < num_sector; i++) {
+		uint32_t sector = start_sector + i;
+		const uint8_t *buffer = data_buf + (i * ctx->page_size);
```
You know that you can avoid multiplication by adding `page_size` on each iteration?

LinderPi committed:
drivers: flash: simulator: Add extended operations

NAND flash translation layers require the flash extended operations API
for bad block management. Add it here without implementation to test an
FTL on the flash simulator.

Signed-off-by: Pascal Linder <pascal.linder@zuehlke.com>

tpambor committed:
tests: drivers: disk: disk_access: add test for ftl disk driver

Add testcase and corresponding overlay for native_sim to test
the FTL (Flash Translation Layer) disk driver.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
modules: dhara: Implement Dhara disk driver as Zephyr module

Zephyr has currently no flash translation layer implemented. This commit
adds the basic possibility to use Dhara as FTL in Zephyr. It is
implemented as disk driver that can be placed between a filesystem and a
flash driver.

Signed-off-by: Pascal Linder <pascal.linder@zuehlke.com>
Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor commented on drivers/disk/ftl_dhara.c:
```diff
@@ -0,0 +1,508 @@
+LOG_MODULE_REGISTER(disk_ftl, CONFIG_DISK_FTL_LOG_LEVEL);
+
+struct disk_ftl_data {
+	struct k_sem lock;
```
I have added a option to configure this. It is ok to leave it disabled e.g. for FATFS with a single volume.

tpambor committed:
drivers: flash: Add bad block management API extensions

Add extended operations to the flash API to support marking blocks
as bad and checking if a block is bad. This is necessary for
NAND flash devices, which can have bad blocks that need to be managed.

Signed-off-by: Pascal Linder <pascal.linder@zuehlke.com>
Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

SuperHeroAbner commented:
Hi @tpambor and @de-nordic , do we have chance to merge this PR recently into v4.4.0? I'm doing [some work](https://github.com/zephyrproject-rtos/zephyr/pull/104870) based on this. Will wait these work to merge first.

tpambor commented:
> Hi @tpambor and @de-nordic , do we have chance to merge this PR recently into v4.4.0? I'm doing [some work](https://github.com/zephyrproject-rtos/zephyr/pull/104870) based on this. Will wait these work to merge first.

I would like to land this as well for v4.4. This PR has been open for a while now. I have addressed all open comments. If there are any remaining issues, I will prioritize fixing them. Downstream, we have been using this for a while without issues.

There are already three drivers with PRs pending that can utilize it:
- https://github.com/zephyrproject-rtos/zephyr/pull/104685 (including in-tree board support for testing)
- https://github.com/zephyrproject-rtos/zephyr/pull/100683 (not planned for v4.4)
- https://github.com/zephyrproject-rtos/zephyr/pull/104870

@de-nordic Please take a look

tpambor committed:
modules: fatfs: support ftl disks when generating FF_VOLUME_STRS

Add disk names for zephyr,ftl-dhara disks to FF_VOLUME_STRS, so that
they are available for mounting.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

de-nordic reviewed (approved):
Looks good. Thanks.
